### ISSUE TITLE
Add MISRA C 2012 Amendment 1 rules supported by PC-Lint

### ISSUE THREAD

**Issue by farshield** [SonarOpenCommunity/sonar-cxx#1856]
There is no definition for MISRA 2012 Rule 21.15, however a similar rule called "L0857" exists.
If I submit the following report to Sonarqube, the issue is ignored and can not be seen in "code smells":

```xml
<?xml version="1.0" ?>
<results>
 <issue file ="mytest.c" line = "254" number = "857" desc = "argument 1 of type &apos;uint8_t *&apos; (aka &apos;unsigned char *&apos;) is not compatible with argument 2 of type &apos;const void *&apos; in call to function &apos;memcpy&apos; [MISRA 2012 Rule 21.15, required]"/>
</results>
```

However, if I remove the text "Rule 21.15"  from the description field of the XML file and submit the report again, I'm able to see the issue in Sonarqube under L0857.

I created a custom rule with the key "M2012_21_15" and then changed the key in Postgresql to "M2012-21.15". I had to do this hack because of input validation when creating the key - I could not use special characters like the dash or the dot so I had to rename it from the database. I labeled it "misrac3" after that.

Once I renamed the key I was able to report the issue and see it in Sonarqube.

Sorry for the offtopic: The best approach would be to add it to the plugin, but in the meantime is there an easier way to rename the key? Especially if I'm using H2 databases (I wasn't able to rename it from H2).

* cxx plugin version: v1.3.2
* SonarQube version: 7.9 LTS

**Comment by guwirth** [comment]
Hello @farshield,

Which tool to create the messages are you using?
Which sensor are you using to read the report?

Regards,

**Comment by farshield** [comment]
Hello @guwirth,

I'm using PC-Lint 1.3 to create the XML report and use the PCLintSensor (`INFO: Sensor C++ (Community) PCLintSensor [cxx] (done) | time=97ms`).

For the sonar-scanner I'm using (v3.2.0.1227-linux), I added the following lines to the "sonar-project.properties file":

```
sonar.projectKey=myproject
sonar.projectName=myproject
sonar.language=c++
sonar.sources=lib
sonar.tests=tests
sonar.c.includeDirectories=lib
sonar.cxx.coverage.reportPath=build/gcovr-report.xml
sonar.cxx.pclint.reportPath=build/pclint-report.xml
```

Regards

**Comment by guwirth** [comment]
Hello @farshield ,

I need some more time to verify this, but maybe you can have a look by yourself:

Code of sensor:
https://github.com/SonarOpenCommunity/sonar-cxx/blob/cxx-1.3.2/cxx-sensors/src/main/java/org/sonar/cxx/sensors/pclint/CxxPCLintSensor.java

Available rules:
https://github.com/SonarOpenCommunity/sonar-cxx/blob/cxx-1.3.2/cxx-sensors/src/main/resources/pclint.xml

Maybe rule is missing? Parsing is wrong...

Regards,

**Comment by guwirth** [comment]
@farshield looking into pclint.xml there is no item `<key>M2012-21.15</key>`. Means the messages are not up to date.

For my understanding new rules of

- MISRA C:2012 Amendment 1
- MISRA C:2012 Addendum 2
- MISRA C:2012 Addendum 3

are missing?

@farshield can you provide these new rules / extend the XML file?
https://github.com/SonarOpenCommunity/sonar-cxx/blob/cxx-1.3.2/cxx-sensors/src/main/resources/pclint.xml

**Comment by farshield** [comment]
@guwirth Indeed this is the case, the rules for MISRA C 2012 AMD-1 are missing. I will try to create first a list of missing rules. Then I will try to add them to the XML file and create a PR if nobody else will create the PR before.

**Comment by guwirth** [comment]
@farshield 

> The following rules supported by PC-Lint are missing from the XML file

Means there are more MISRA C 2012 rules missing in the XML but these are not supported by PC-Lint?

**Comment by farshield** [comment]
@guwirth 

Correct. Here is a list of MISRA C 2012 AMD-1 supported rules found in the documentation of the latest PC-Lint Plus (v1.3.5-beta):

![image](https://user-images.githubusercontent.com/14978226/80801335-4cf54680-8bac-11ea-9faa-82d96b83dfc5.png)

**Comment by guwirth** [comment]
closed with #1857
